packages/theming-core/src/generics/themeRegistry.ts

```diff
@@ -0,0 +1,247 @@
+/**
+ * Function which takes a parent theme, does some processing on it, and returns a new theme definition.
+ * This definition will then be passed to the standard resolver
+ *
+ * @internal This is an experimental interface and will be changed post design review.
+ */
+export type ProcessTheme<ITheme, IThemeDefinition> = (parentTheme: ITheme) => IThemeDefinition;
+
+/**
+ * Function which takes a theme definition and a parent theme and produces a new resolved theme
+ *
+ * @internal This is an experimental interface and will be changed post design review.
+ */
+export type ResolveTheme<ITheme, IThemeDefinition> = (definition: IThemeDefinition | undefined, parentTheme: ITheme) => ITheme;
+
+/**
+ * Graph node in teh theme graph
+ */
+interface IThemeEntry<ITheme, IThemeDefinition> {
+  /** parent theme */
+  parent?: string;
+
+  /** definition or function used to regenerate the theme */
+  definition?: IThemeDefinition | ProcessTheme<ITheme, IThemeDefinition>;
+
+  /** resolved theme */
+  resolved?: ITheme;
+}
+
+/** Predefined entries in the theme graph, pretty self-explanatory */
+interface IThemeGraphCore<ITheme, IThemeDefinition> {
+  _platform: IThemeEntry<ITheme, IThemeDefinition>;
+  default: IThemeEntry<ITheme, IThemeDefinition>;
+}
+
+/** Key-names for the platform and default themes */
+const _platformKey: keyof IThemeGraphCore<object, object> = '_platform';
+const _defaultKey: keyof IThemeGraphCore<object, object> = 'default';
+
+/** Graph interface, takes the predefined entries and adds the index signature */
+interface IThemeGraph<ITheme, IThemeDefinition> extends IThemeGraphCore<ITheme, IThemeDefinition> {
+  [key: string]: IThemeEntry<ITheme, IThemeDefinition>;
+}
+
+/**
+ * Function signature for theme registration
+ * @internal This is an experimental interface and will be changed post design review.
+ */
+export type IRegisterTheme<ITheme, IThemeDefinition> =
+  (definition: IThemeDefinition | ProcessTheme<ITheme, IThemeDefinition>, name?: string, parent?: string) => void;
+
+/**
+ * Function signature for get theme calls
+ * @internal This is an experimental interface and will be changed post design review.
```

Comment:
nit:

We just started trying to be more consistent in our comments across Fabric. One of the things we are encouraging is for function comments to imply an action with a verb. Specifically for getters, the comment should start with "Gets".

In this case, for example, it would be:
```suggestion
 * Gets theme calls.
 * @internal This is an experimental interface and will be changed post design review.
```

For more info on commenting conventions you can visit [here](https://github.com/OfficeDev/office-ui-fabric-react/wiki/Commenting-conventions).